File: src/TaskOTime/TaskOTime.ViewModel/ViewModels/TaskItemViewModel.vb
Imports TaskOTime.ViewModel.Base

Namespace ViewModels
    Public Class TaskItemViewModel
        Inherits ViewModelBase

        Private ReadOnly _id As Guid
        Private _title As String
        Private _description As String
        Private _dueText As String
        Private _isStarted As Boolean
        Private _isDone As Boolean
        Private _needsMore As Boolean
        Private _startedAt As DateTime?
        Private _recordingElapsed As TimeSpan

        Public Sub New(title As String, description As String, dueText As String)
            Me.New(Guid.NewGuid(), title, description, dueText)
        End Sub

        Public Sub New(id As Guid, title As String, description As String, dueText As String)
            _id = id
            _title = title
            _description = description
            _dueText = dueText
        End Sub

        Public ReadOnly Property Id As Guid
            Get
                Return _id
            End Get
        End Property

        Public Property Title As String
            Get
                Return _title
            End Get
            Set(value As String)
                SetProperty(_title, value, NameOf(Title))
            End Set
        End Property

        Public Property Description As String
            Get
                Return _description
            End Get
            Set(value As String)
                SetProperty(_description, value, NameOf(Description))
            End Set
        End Property

        Public Property DueText As String
            Get
                Return _dueText
            End Get
            Set(value As String)
                SetProperty(_dueText, value, NameOf(DueText))
            End Set
        End Property

        Public Property IsStarted As Boolean
            Get
                Return _isStarted
            End Get
            Private Set(value As Boolean)
                If SetProperty(_isStarted, value, NameOf(IsStarted)) Then
                    RaiseStatusPropertiesChanged()
                End If
            End Set
        End Property

        Public ReadOnly Property StartedAt As DateTime?
            Get
                Return _startedAt
            End Get
        End Property

        Public Property RecordingElapsed As TimeSpan
            Get
                Return _recordingElapsed
            End Get
            Private Set(value As TimeSpan)
                If SetProperty(_recordingElapsed, value, NameOf(RecordingElapsed)) Then
                    OnPropertyChanged(NameOf(RecordingElapsedText))
                    OnPropertyChanged(NameOf(RecordingStatusText))
                End If
            End Set
        End Property

        Public Property IsDone As Boolean
            Get
                Return _isDone
            End Get
            Private Set(value As Boolean)
                If SetProperty(_isDone, value, NameOf(IsDone)) Then
                    RaiseStatusPropertiesChanged()
                End If
            End Set
        End Property

        Public Property NeedsMore As Boolean
            Get
                Return _needsMore
            End Get
            Private Set(value As Boolean)
                If SetProperty(_needsMore, value, NameOf(NeedsMore)) Then
                    RaiseStatusPropertiesChanged()
                End If
            End Set
        End Property

        Public ReadOnly Property IsOpen As Boolean
            Get
                Return Not IsDone
            End Get
        End Property

        Public ReadOnly Property StatusText As String
            Get
                If IsDone Then
                    Return "Erledigt"
                End If

                If NeedsMore Then
                    Return "Mehr zu tun"
                End If

                If IsStarted Then
                    Return "Läuft"
                End If

                Return "Bereit"
            End Get
        End Property

        Public ReadOnly Property ShortTitle As String
            Get
                If String.IsNullOrEmpty(Title) OrElse Title.Length <= 20 Then
                    Return Title
                End If

                Return Title.Substring(0, 20)
            End Get
        End Property

        Public ReadOnly Property RecordingElapsedText As String
            Get
                Return $"{CInt(Math.Floor(RecordingElapsed.TotalHours)):00}:{RecordingElapsed.Minutes:00}:{RecordingElapsed.Seconds:00}"
            End Get
        End Property

        Public ReadOnly Property RecordingStatusText As String
            Get
                If Not IsStarted OrElse Not StartedAt.HasValue Then
                    Return String.Empty
                End If

                Return $"{ShortTitle} seit {StartedAt.Value:HH:mm} - {RecordingElapsedText}"
            End Get
        End Property

        Public ReadOnly Property ActionHint As String
            Get
                If IsDone Then
                    Return "Diese Aufgabe ist abgeschlossen."
                End If

                If NeedsMore Then
                    Return "Die Aufgabe bleibt offen und braucht weitere Schritte."
                End If

                If IsStarted Then
                    Return "Die Aufgabe läuft gerade."
                End If

                Return "Bereit zum Starten."
            End Get
        End Property

        Public Sub MarkStarted(Optional startedAt As DateTime? = Nothing)
            IsDone = False
            NeedsMore = False
            _startedAt = If(startedAt, DateTime.Now)
            OnPropertyChanged(NameOf(StartedAt))
            IsStarted = True
            UpdateRecordingElapsed(DateTime.Now)
        End Sub

        Public Sub MarkDone()
            NeedsMore = False
            _startedAt = Nothing
            RecordingElapsed = TimeSpan.Zero
            IsStarted = False
            IsDone = True
        End Sub

        Public Sub MarkNeedsMore()
            IsDone = False
            _startedAt = Nothing
            RecordingElapsed = TimeSpan.Zero
            IsStarted = False
            NeedsMore = True
        End Sub

        Public Sub StopRecordingWithoutFinishing()
            If Not IsStarted Then
                Return
            End If

            _startedAt = Nothing
            RecordingElapsed = TimeSpan.Zero
            IsStarted = False
        End Sub

        Public Sub UpdateRecordingElapsed(nowValue As DateTime)
            If Not IsStarted OrElse Not StartedAt.HasValue Then
                RecordingElapsed = TimeSpan.Zero
                Return
            End If

            RecordingElapsed = nowValue - StartedAt.Value
        End Sub

        Private Sub RaiseStatusPropertiesChanged()
            OnPropertyChanged(NameOf(IsOpen))
            OnPropertyChanged(NameOf(StatusText))
            OnPropertyChanged(NameOf(ActionHint))
            OnPropertyChanged(NameOf(ShortTitle))
            OnPropertyChanged(NameOf(RecordingStatusText))
        End Sub
    End Class
End Namespace

File: src/TaskOTime/TaskOTime.ViewModel/ViewModels/TaskItemViewModel.cs
using System;
using TaskOTime.ViewModel.Base;

namespace TaskOTime.ViewModel.ViewModels
{
    public class TaskItemViewModel : Localization.LocalizedViewModelBase
    {

        private readonly Guid _id;
        public Guid IdProject { get; set; }
        public Guid IdTask { get; set; }
        private string _title;
        private string _description;
        private string _dueText;
        private bool _isStarted;
        private bool _isDone;
        private bool _needsMore;
        private DateTime? _startedAt;
        private TimeSpan _recordingElapsed;

        public TaskItemViewModel(string title, string description, string dueText) : this(Guid.NewGuid(), title, description, dueText)
        {
        }

        public TaskItemViewModel(Guid id, string title, string description, string dueText)
        {
            _id = id;
            _title = title;
            _description = description;
            _dueText = dueText;
        }

        public Guid Id
        {
            get
            {
                return _id;
            }
        }

        public string Title
        {
            get
            {
                return _title;
            }
            set
            {
                SetProperty(ref _title, value, nameof(Title));
            }
        }

        public string Description
        {
            get
            {
                return _description;
            }
            set
            {
                SetProperty(ref _description, value, nameof(Description));
            }
        }

        public string DueText
        {
            get
            {
                return _dueText;
            }
            set
            {
                SetProperty(ref _dueText, value, nameof(DueText));
            }
        }

        public bool IsStarted
        {
            get
            {
                return _isStarted;
            }
            private set
            {
                if (SetProperty(ref _isStarted, value, nameof(IsStarted)))
                {
                    RaiseStatusPropertiesChanged();
                }
            }
        }

        public DateTime? StartedAt
        {
            get
            {
                return _startedAt;
            }
        }

        public TimeSpan RecordingElapsed
        {
            get
            {
                return _recordingElapsed;
            }
            private set
            {
                if (SetProperty(ref _recordingElapsed, value, nameof(RecordingElapsed)))
                {
                    OnPropertyChanged(nameof(RecordingElapsedText));
                    OnPropertyChanged(nameof(RecordingStatusText));
                }
            }
        }

        public bool IsDone
        {
            get
            {
                return _isDone;
            }
            private set
            {
                if (SetProperty(ref _isDone, value, nameof(IsDone)))
                {
                    RaiseStatusPropertiesChanged();
                }
            }
        }

        public bool NeedsMore
        {
            get
            {
                return _needsMore;
            }
            private set
            {
                if (SetProperty(ref _needsMore, value, nameof(NeedsMore)))
                {
                    RaiseStatusPropertiesChanged();
                }
            }
        }

        public bool IsOpen
        {
            get
            {
                return !IsDone;
            }
        }

        public string StatusText
        {
            get
            {
                if (IsDone)
                {
                    return Text("Task_Done");
                }

                if (NeedsMore)
                {
                    return Text("Task_More");
                }

                if (IsStarted)
                {
                    return Text("Task_Running");
                }

                return Text("Task_Ready");
            }
        }

        public string ShortTitle
        {
            get
            {
                if (string.IsNullOrEmpty(Title) || Title.Length <= 20)
                {
                    return Title;
                }

                return Title.Substring(0, 20);
            }
        }

        public string RecordingElapsedText
        {
            get
            {
                return $"{(int)Math.Round(Math.Floor(RecordingElapsed.TotalHours)):00}:{RecordingElapsed.Minutes:00}:{RecordingElapsed.Seconds:00}";
            }
        }

        public string RecordingStatusText
        {
            get
            {
                if (!IsStarted || !StartedAt.HasValue)
                {
                    return string.Empty;
                }

                return Text("Task_Recording", ShortTitle, StartedAt.Value, RecordingElapsedText);
            }
        }

        public string ActionHint
        {
            get
            {
                if (IsDone)
                {
                    return Text("Task_DoneHint");
                }

                if (NeedsMore)
                {
                    return Text("Task_MoreHint");
                }

                if (IsStarted)
                {
                    return Text("Task_RunningHint");
                }

                return Text("Task_ReadyHint");
            }
        }

        public void MarkStarted(DateTime? startedAt = default)
        {
            IsDone = false;
            NeedsMore = false;
            _startedAt = startedAt ?? DateTime.Now;
            OnPropertyChanged(nameof(startedAt));
            IsStarted = true;
            UpdateRecordingElapsed(DateTime.Now);
        }

        internal TaskItemViewModel CreateCompletedSnapshot()
        {
            var completed = new TaskItemViewModel(Id, Title, Description, DueText)
            {
                IdProject = IdProject,
                IdTask = IdTask
            };
            completed.MarkDone();
            return completed;
        }

        public void MarkDone()
        {
            NeedsMore = false;
            _startedAt = default;
            RecordingElapsed = TimeSpan.Zero;
            IsStarted = false;
            IsDone = true;
        }

        public void MarkNeedsMore()
        {
            IsDone = false;
            _startedAt = default;
            RecordingElapsed = TimeSpan.Zero;
            IsStarted = false;
            NeedsMore = true;
        }

        public void StopRecordingWithoutFinishing()
        {
            if (!IsStarted)
            {
                return;
            }

            _startedAt = default;
            RecordingElapsed = TimeSpan.Zero;
            IsStarted = false;
        }

        public void UpdateRecordingElapsed(DateTime nowValue)
        {
            if (!IsStarted || !StartedAt.HasValue)
            {
                RecordingElapsed = TimeSpan.Zero;
                return;
            }

            RecordingElapsed = nowValue - StartedAt.Value;
        }

        private void RaiseStatusPropertiesChanged()
        {
            OnPropertyChanged(nameof(IsOpen));
            OnPropertyChanged(nameof(StatusText));
            OnPropertyChanged(nameof(ActionHint));
            OnPropertyChanged(nameof(ShortTitle));
            OnPropertyChanged(nameof(RecordingStatusText));
        }
    }
}